### PR TITLE
(#3464) fix Query failed when fact table has no data in join case

### PR DESCRIPTION
**Pull request by wangbo**
#3464

major work
1.  Correct the value of ```numNodes``` and ```cardinality``` when ```OlapTableScan``` computeStats so that the ``` broadcast cost``` and ```paritition join cost ``` can be calculated correctly.
2. Find a input fragment with higher parallelism for shuffle fragment to assign backend

### CODE REVIEW

**Review by morningman — DISMISSED** [review]
LGTM

File: fe/src/main/java/org/apache/doris/qe/Coordinator.java [hunk 1]
```diff
@@ -914,17 +914,29 @@ private void computeFragmentHosts() throws Exception {
                 // there is no leftmost scan; we assign the same hosts as those of our
                 // leftmost input fragment (so that a partitioned aggregation
                 // fragment runs on the hosts that provide the input data)
```
Comment (kangkaisen) [inline fe/src/main/java/org/apache/doris/qe/Coordinator.java:916]: Please also update the comment.

File: fe/src/main/java/org/apache/doris/qe/Coordinator.java [hunk 2]
```diff
@@ -914,17 +914,29 @@ private void computeFragmentHosts() throws Exception {
...
-                PlanFragmentId inputFragmentIdx = fragments.get(i).getChild(0).getFragmentId();
+
+                // find a input fragment which has a higher parallelism
+                int inputFragmentIndex = 0;
+                int maxParallelism = 0;
```
Comment (kangkaisen) [inline fe/src/main/java/org/apache/doris/qe/Coordinator.java:920]: I think `parallelism` rename  to  instance_number is better.

Comment (wangbo) [reply fe/src/main/java/org/apache/doris/qe/Coordinator.java:920]: 👌

File: fe/src/main/java/org/apache/doris/qe/Coordinator.java (hunk 1 above)
Comment (wangbo) [reply fe/src/main/java/org/apache/doris/qe/Coordinator.java:916]: 👌

**Comment by wangbo** [comment]
After I run this pr in out production env,I find a new problem.

A dim table whose size is 1.1 GB with single replica, and 1.1G is less than perNodeMemLimit(2GB)，so when the fact table is empty the broadcast join is choosen.
But when the hash table actually builds, the memory usesage still execeeds which means HASH_TBL_SPACE_OVERHEAD(=1.1) is not so precise.

This pr can fix  the is problem when user use ```shuffle hint``` because shuffle Node runs more nodes.

So I think there may two ways to solve the problem thoroughly;
1  make HASH_TBL_SPACE_OVERHEAD a configurable  param,but it seems not a common solution, because I'm not sure what factors can affect the real hash table size.
2 maybe we can disable broadcast when dim table exceeds 100M

**Comment by morningman** [comment]
> 1 make HASH_TBL_SPACE_OVERHEAD a configurable param,but it seems not a common solution, because I'm not sure what factors can affect the real hash table size.
> 2 maybe we can disable broadcast when dim table exceeds 100M

These 2 solutions looks seem, that both require manual parameter adjustment. This is no different from directly adding hints manually.

**Comment by wangbo** [comment]
@morningman 
The key point is that it's currently impossible to accurately estimate the size of the memory footprint at runtime according to table size.Otherwise, we can choose the right join mode or the correct num backends.
At present, it seems that it can only be adjusted manually. User decide the join mode.

**Comment by morningman** [comment]
> @morningman
> The key point is that it's currently impossible to accurately estimate the size of the memory footprint at runtime according to table size.Otherwise, we can choose the right join mode or the correct num backends.
> At present, it seems that it can only be adjusted manually. User decide the join mode.

So I think just use the `shuffle hint` is enough? No need to add new config `HASH_TBL_SPACE_OVERHEAD` or disallow broadcast large than 100MB?

**Comment by wangbo** [comment]
> > @morningman
> > The key point is that it's currently impossible to accurately estimate the size of the memory footprint at runtime according to table size.Otherwise, we can choose the right join mode or the correct num backends.
> > At present, it seems that it can only be adjusted manually. User decide the join mode.
> 
> So I think just use the `shuffle hint` is enough? No need to add new config `HASH_TBL_SPACE_OVERHEAD` or disallow broadcast large than 100MB?

I think we can set ```partition join``` as default  when we're unable to estimate the cost accurately and broadcast join cost equals partition join cost and without user hint.
From
```
                && broadcastCost <= partitionCost))) {
            doBroadcast = true;
        } else {
            doBroadcast = false;
        }
```
to
```
                && broadcastCost < partitionCost))) {
            doBroadcast = true;
        } else {
            doBroadcast = false;
        }
```

**Comment by morningman** [comment]
> I think we can set `partition join` as default when we're unable to estimate the cost accurately and broadcast join cost equals partition join cost and without user hint.

This LGTM.

**Comment by wangbo** [comment]
set partition join as default when partition join cost equals broadcast join

**Review by morningman — APPROVED** [review]
LGTM

**Review by kangkaisen — APPROVED** [review]
+1， LGTM

**Pull request merged**